src/D2L.CodeStyle.Analyzers/Immutability/ImmutableTypeInfo.cs

```diff
@@ -1,34 +1,76 @@
 ﻿using System;
+using System.Collections.Immutable;
+using System.Linq;
 using Microsoft.CodeAnalysis;
 
 namespace D2L.CodeStyle.Analyzers.Immutability {
 	/// <summary>
 	/// A representation of the immutability of some type.
 	/// </summary>
 	internal readonly struct ImmutableTypeInfo : IEquatable<ImmutableTypeInfo> {
-		public ImmutableTypeInfo(
+
+		private ImmutableTypeInfo(
 			ImmutableTypeKind kind,
-			ITypeSymbol type
+			INamedTypeSymbol type,
+			ImmutableArray<bool> immutableTypeParameters

```

Comment:
Maybe add a comment somewhere to explain how this works? I assume `[Immutable]` -> `true` here?

Reply:
Yes from https://github.com/Brightspace/D2L.CodeStyle/pull/595/files#diff-e61de6c1c372a9fe62f866659862da11c2f471672e110100560ab084e4183ccdR34

doc comment on `public ImmutableArray<bool> ImmutableTypeParameters { get; }`?

Reply:
🤷 dno

Reply:
Hmmm.... I wonder if some of the logic could move into `ImmutableTypeInfo` and more of it could just become private?

That makes me thing.... `m_extraImmutableTypes`... the keys, should they be the generic type instead?

Reply:
```cs
public sealed class ImmutableTypeInfo {
  INamedType Type { get; }
  ImmutableType Kind { get; }

  private ImmutableTypeInfo( ... ) {
  }

  // your two static helper Create- methods

  public bool IsImmutable( ImmutabilityContext ctx, INamedType other ) {
    // say we are ImmutableArray<[Immutable]T>
    // other is ImmutableArray<int>
    // we need to check that T=int is immutable
  }
}
```

that's what I mean

Reply:
> That makes me thing.... `m_extraImmutableTypes`... the keys, should they be the generic type instead?

I think we'll need both. We want `ImmutableArray<T>` to be considered immutable in general and so yeah we'd use the OriginalDefinition as a key, but maybe also `IFactory<IMyFoo>` (for replacing `ImmutableGeneric`)?

Reply:
That way there is no need to document stuff. `ImmutableTypeInfo.IsImmutable`/`ImmutabilityContext.IsImmutable` would be co-recursive to support `ImmutableArray<ImmutableArray<int>>` etc.

Reply:
Ah `OriginalDefinition`, ok now I get your other comment.

Reply:
Yeah, only concern with the co-recursiveness for me would be splitting the logic up too terribly, but some of it is via the `Create` and knowing how to use the properties anyway, so :shrug: seems good